### PR TITLE
Adding Upsampling 3D layer and tests for upsampling layers

### PR DESCRIPTION
**Pull request by Shashi456**

### CODE REVIEW

File: Sources/DeepLearning/Layer.swift [hunk 1]
```diff
@@ -1169,6 +1169,33 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    /// Creates an upsampling layer.
+    ///
+    /// - Parameter size: The upsampling factor for rows and columns.
+    public init(size: Int) {
+       self.size = size
+    }
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameters:
```
Comment (dan-zheng) [inline Sources/DeepLearning/Layer.swift]: `Layer.swift` isn't consistent, but could you please use:
```suggestion
    /// - Parameter input: The input to the layer.
```

**Comment by Shashi456** [comment]
@dan-zheng should i add tests for the other upsampling layers in this PR?

**Comment by dan-zheng** [comment]
> @dan-zheng should i add tests for the other upsampling layers in this PR?

Yes please, that would be splendid!

File: Tests/DeepLearningTests/LayerTests.swift [hunk 2]
```diff
@@ -66,6 +66,15 @@ final class LayerTests: XCTestCase {
         XCTAssertEqual(output, expected)
     }
 
+    func TestUpSampling3D(){
```
Comment (dan-zheng) [inline Tests/DeepLearningTests/LayerTests.swift]: Minor: please use camelcase and add a space.
```suggestion
    func testUpSampling3D() {
```

Comment (Shashi456) [reply Tests/DeepLearningTests/LayerTests.swift]: @dan-zheng Is the logic right? I'm sorta unsure about the assertion i'm doing.

Comment (Shashi456) [reply Tests/DeepLearningTests/LayerTests.swift]: should `expected = TensorShape([1, shape[0]* size, ., . ,1])`

File: Sources/DeepLearning/Layer.swift [hunk 3]
```diff
@@ -1169,6 +1169,33 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    ///
+    /// - Parameter size: The upsampling factor for rows and columns.
+    public init(size: Int) {
+       self.size = size
+    }
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    ///
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: I'd suggest removing this empty line.

File: Sources/DeepLearning/Layer.swift [hunk 4]
```diff
@@ -1169,6 +1169,34 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The output.
+    @differentiable
+    public func call(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let shape = input.shape
+        let (batchSize, height, width, breadth, channels) = (shape[0], shape[1], shape[2]
+                                                            , shape[3], shape[4])
+        let scaleOnes = Tensor<Scalar>(ones: [1, 1, size, 1, size, 1, size, 1])
+        let upSampling = input.reshaped(to: [batchSize, height, 1, width, 1,
+                                            breadth, 1, channels]) * scaleOnes
```
Comment (tanmayb123) [inline Sources/DeepLearning/Layer.swift]: small question: `breadth` or `depth`?

File: Sources/DeepLearning/Layer.swift [hunk 5]
```diff
@@ -1169,6 +1169,34 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The output.
+    @differentiable
+    public func call(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let shape = input.shape
+        let (batchSize, height, width, breadth, channels) = (shape[0], shape[1], shape[2]
+                                                            , shape[3], shape[4])
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: When breaking a tuple to multiple lines, it's better to align elements that are not on the first line with the first element and not to make commas start a line, roughly:
```swift
(shape[0], shape[1], shape[2],
 shape[3], shape[4])
```

Another option is to put the entire tuple on the second line:

```swift
        let (batchSize, height, width, breadth, channels) =
            (shape[0], shape[1], shape[2], shape[3], shape[4])
```

What do you think?

File: Sources/DeepLearning/Layer.swift [hunk 6]
```diff
@@ -1169,6 +1169,34 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+        let scaleOnes = Tensor<Scalar>(ones: [1, 1, size, 1, size, 1, size, 1])
+        let upSampling = input.reshaped(to: [batchSize, height, 1, width, 1,
+                                            breadth, 1, channels]) * scaleOnes
+        return upSampling.reshaped(to: [batchSize, height * size, width * size, breadth * size, channels])
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Could you make sure this fits within 100 columns?

File: Tests/DeepLearningTests/LayerTests.swift [hunk 7]
```diff
@@ -66,6 +66,33 @@ final class LayerTests: XCTestCase {
...
+      let expected = TensorShape([1, input.shape[1] * size, input.shape[2] * size, 1])
+      XCTAssertEqual(output.shape, expected)
+    }
+
+    func testUpSampling3D() {
+      let size = 6
+      let layer = UpSampling3D<Float>(size: size)
+      let input = Tensor<Float>(shape: [1, 4, 3, 2, 1], scalars: (0..<24).map(Float.init))
+      let output = layer.inferring(from: input)
+      let expected = TensorShape([1, input.shape[1] * size, input.shape[2] * size, input.shape[3] * size, 1])
```
Comment (rxwei) [inline Tests/DeepLearningTests/LayerTests.swift]: Fit within 100 columns as well.

File: Sources/DeepLearning/Layer.swift (hunk 4 above)
Comment (Shashi456) [reply Sources/DeepLearning/Layer.swift]: Actually depth and height would be similar. i think it should be length, width and height, i took your suggestion for now. Thank you.

File: Sources/DeepLearning/Layer.swift [hunk 8]
```diff
@@ -1169,6 +1169,35 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The output.
+    @differentiable
+    public func call(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let shape = input.shape
+        let (batchSize, height, width, depth, channels) =
+                                      (shape[0], shape[1], shape[2], shape[3], shape[4])
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Thanks for the change! Just a minor suggestion: This line should be indented by 4 from `let`.

File: Sources/DeepLearning/Layer.swift [hunk 9]
```diff
@@ -1169,6 +1169,35 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+        let scaleOnes = Tensor<Scalar>(ones: [1, 1, size, 1, size, 1, size, 1])
+        let upSampling = input.reshaped(
+        to: [batchSize, height, 1, width, 1, depth, 1, channels]) * scaleOnes
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Could you please indent this line by 4 also?

File: Sources/DeepLearning/Layer.swift [hunk 10]
```diff
@@ -1169,6 +1169,35 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+        return upSampling.reshaped(
+        to: [batchSize, height * size, width * size, depth * size, channels])
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: And this line. Thanks!

File: Sources/DeepLearning/Layer.swift (hunk 8 above)
Comment (Shashi456) [reply Sources/DeepLearning/Layer.swift]: oh alright, I'm sorry!

File: Sources/DeepLearning/Layer.swift [hunk 11]
```diff
@@ -1169,6 +1169,35 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+
+    /// Returns the output obtained from applying the layer to the given input.
+    ///
+    /// - Parameter input: The input to the layer.
+    /// - Returns: The output.
+    @differentiable
+    public func call(_ input: Tensor<Scalar>) -> Tensor<Scalar> {
+        let shape = input.shape
+        let (batchSize, height, width, depth, channels) =
+                (shape[0], shape[1], shape[2], shape[3], shape[4])
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]: Could you please make sure they are indented by 4 spaces from `let`?
```suggestion
            (shape[0], shape[1], shape[2], shape[3], shape[4])
```

File: Sources/DeepLearning/Layer.swift [hunk 12]
```diff
@@ -1169,6 +1169,35 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+        let scaleOnes = Tensor<Scalar>(ones: [1, 1, size, 1, size, 1, size, 1])
+        let upSampling = input.reshaped(
+                to: [batchSize, height, 1, width, 1, depth, 1, channels]) * scaleOnes
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]:
```suggestion
            to: [batchSize, height, 1, width, 1, depth, 1, channels]) * scaleOnes
```

File: Sources/DeepLearning/Layer.swift [hunk 13]
```diff
@@ -1169,6 +1169,35 @@ public struct UpSampling2D<Scalar: TensorFlowFloatingPoint>: Layer {
...
+        return upSampling.reshaped(
+                to: [batchSize, height * size, width * size, depth * size, channels])
```
Comment (rxwei) [inline Sources/DeepLearning/Layer.swift]:
```suggestion
            to: [batchSize, height * size, width * size, depth * size, channels])
```

File: Sources/DeepLearning/Layer.swift (hunk 11 above)
Comment (Shashi456) [reply Sources/DeepLearning/Layer.swift]: Oh I made them 4 tabs instead of spaces. My bad.

**Comment by Shashi456** [comment]
@rxwei I think this can be merged.

**Comment by rxwei** [comment]
Thanks!

**Comment by dan-zheng** [comment]
Note: the test for `UpSampling3D` is broken and [`testUpSampling3D` has been disabled](https://github.com/tensorflow/swift-apis/pull/130/files#diff-f511a02d8e443c3804257b286af03945R137).
Please see [TF-525](https://bugs.swift.org/browse/TF-525) for more details.

**Pull request merged**